lightning/src/events/mod.rs

```diff
@@ -465,6 +466,25 @@ impl_writeable_tlv_based_enum_upgradable!(ClosureReason,
 	},
 );
 
+/// The reason for HTLC failures in [`HTLCDestination`].
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum HTLCDestinationFailure {
+	/// The forwarded HTLC was failed back by the downstream node with an encrypted error reason.
+	Downstream,
+	/// The HTLC was failed locally by our node.
+	Local{
+		/// The reason that our node chose to fail the HTLC.
+		reason: LocalHTLCFailureReason
```

Comment:
If the goal is to expose all information about a failure to the user, shouldn't the failure parameters be included here too? And possibly the tlv extensions?

Perhaps not in this PR, but just thinking ahead whether it would work.

Reply:
I think we'd probably put that information in `HTLCHandlingFailed` because it applies across all types of `HTLCDestination` types?

Reply:
I think that's true, except for `HTLCDestinationFailure::Downstream` ?

Reply:
Clarified offline. Can extend the `LocalHTLCFailureReason enum` with the failure type specific parameters and tlv extensions.

Reply:
Discussed a bit offline: thinking about the case where we want to add extra information to these errors if there's demand for it (eg, adding `htlc_amount` to the `FeeInsufficient` variant).

For `LocalHTLCFailureReason`, we an always add fields to the enum struct. Likewise for `Downstream`, could add a field because we're using `impl_writeable_tlv_based_enum`, but are unlikely to need it as error reasons are encrypted.